test/object-store/realm.cpp

```diff
@@ -1206,6 +1207,117 @@ TEST_CASE("Get Realm using Async Open", "[sync][pbs][async open]") {
         REQUIRE(got_error);
     }
 
+    SECTION("waiters are cancelled if cancel_waits_on_nonfatal_error") {
+        auto logger = util::Logger::get_default_logger();
+        auto transport = std::make_shared<HookedUnitTestTransport>();
+        auto socket_provider = std::make_shared<HookedSocketProvider>(logger, "some user agent");
+        enum TestMode { location_fails, token_fails, token_not_authorized };
+        auto txt_test_mode = [](TestMode mode) {
+            switch (mode) {
+                case TestMode::location_fails:
+                    return "location_fails";
+                case TestMode::token_fails:
+                    return "token_fails";
+                case TestMode::token_not_authorized:
+                    return "token_not_authorized";
+                default:
+                    return "Unknown TestMode";
+            }
+        };
+
+        OfflineAppSession::Config oas_config;
+        oas_config.transport = transport;
+        oas_config.socket_provider = socket_provider;
+        OfflineAppSession oas(oas_config);
+
+        SyncTestFile config(oas, "realm");
+        auto user = config.sync_config->user;
+        config.sync_config->cancel_waits_on_nonfatal_error = true;
+        config.sync_config->error_handler = [&logger](std::shared_ptr<SyncSession> session, SyncError error) {
+            logger->debug("The sync error handler caught an error: '%1' for '%2'", error.status, session->path());
+            // Ignore connection failed non-fatal errors and check for access token refresh unauthorized fatal errors
+            if (error.status.code() == ErrorCodes::SyncConnectFailed) {
+                REQUIRE_FALSE(error.is_fatal);
+                return;
+            }
+            if (error.status.code() == ErrorCodes::AuthError) {
+                REQUIRE(error.is_fatal);
+            }
+        };
+
+        // User should be logged in at this point
+        REQUIRE(user->is_logged_in());
+        bool not_authorized = false;
+        bool token_refresh_called = false;
+        bool location_refresh_called = false;
+
+        TestMode mode = GENERATE(location_fails, token_fails, token_not_authorized);
+
+        SECTION(util::format("access token expired when realm is opened - mode: %1", txt_test_mode(mode))) {
```

Comment:
Catch has DYNAMIC_SECTION for this: `DYNAMIC_SECTION("access token expired when realm is opened - mode: " << txt_test_mode(mode))`.

This pattern of doing configuration in the SECTION and then the shared test afterwards has the problem that it results in catch not being able to tell us which section failed, and relying on logging for that is awkward. I think the intended way to do this with Catch would be using GENERATE on a second enum and switching on that instead of using sections. `CAPTURE(mode)` can be used to make it print the generated mode on failure.

Reply:
good to know 👍

Reply:
Thanks @tgoyne - I updated the tests to use the `DYNAMIC_SECTION` so the individual sections would show up in the test results.